balancer/rls/balancer_test.go

```diff
@@ -650,6 +650,180 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+            ]
+        },
+        {
+            "key": "k2",
+            "names": [
+                "n2"
+            ]
+        }
+    ]
+    `
+
+	configJSON := `
+	{
+	  "loadBalancingConfig": [
+		{
+		  "%s": {
+			"routeLookupConfig": {
+				"grpcKeybuilders": [{
+					"names": [{"service": "grpc.testing.TestService"}],
+					"headers": %s
+				}],
+				"lookupService": "%s",
+				"cacheSizeBytes": %d
+			},
+			"childPolicy": [{"%s": {}}],
+			"childPolicyConfigTargetFieldName": "Backend"
+		  }
+		}
+	  ]
+	}`
+	scJSON := fmt.Sprintf(configJSON, topLevelBalancerName, headers, rlsServer.Address, 1000, childPolicyName)
+	sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON)
+	r.InitialState(resolver.State{ServiceConfig: sc})
+
+	ccCh := make(chan *grpc.ClientConn, 1)
+	go func() {
+		cc, err := grpc.Dial(r.Scheme()+":///", grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return
+		}
+		ccCh <- cc
+	}()
+	<-clientConnUpdateDone
+	cc := <-ccCh
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	makeTestRPCAndVerifyError(ctx, t, cc, codes.Unavailable, nil)
+	t.Logf("Verifying if RPC failed when listener is stopped.")
+
+	ctxOutgoing := metadata.AppendToOutgoingContext(ctx, "n1", "v1")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh1)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	ctxOutgoing = metadata.AppendToOutgoingContext(ctx, "n2", "v2")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh2)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	// Override the newPickerGenerated to measure the number of times
+	// the picker is generated because state updates can be inhibited.
+	//
+	// Note: This needs to be initialised here and not in the beginning
+	// of the test, as otherwise for every RPC call we make, it'd have
+	// affected total picker count.
+	pckrSentBeforeClientConnUpdate := make(chan struct{}, 1)
+	totalPickerGenerated := 0
+	origNewPickerGenerated := newPickerGenerated
+	newPickerGenerated = func() {
+		totalPickerGenerated++
+		if totalPickerGenerated == 2 {
+			pckrSentBeforeClientConnUpdate <- struct{}{}
+		}
+	}
+	defer func() { newPickerGenerated = origNewPickerGenerated }()
+
+	// Setting the size to 1 will cause the entries to be evicted.
+	scJSON1 := fmt.Sprintf(configJSON, topLevelBalancerName, headers, rlsServer.Address, 1, childPolicyName)
+	sc1 := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON1)
+	go r.UpdateState(resolver.State{ServiceConfig: sc1})
+	select {
+	case <-pckrSentBeforeClientConnUpdate:
+	case <-clientConnUpdateDone:
+		t.Fatalf("Client conn update was completed before picker update.")
+	case <-ctx.Done():
+		t.Errorf("client conn update could not complete: %v", ctx.Err().Error())
+	}
+
+	// Once picker was updated, wait for client conn update
+	// to complete.
+	<-clientConnUpdateDone
```

Comment:
Its probably a good idea to wait for context timeout along with this to ensure the test doesn't hang for a long time in case of failures.